src/hotspot/share/runtime/methodUnloadBlocker.inline.hpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_METHOD_UNLOAD_BLOCKER_HANDLE_INLINE_HPP
+#define SHARE_RUNTIME_METHOD_UNLOAD_BLOCKER_HANDLE_INLINE_HPP
+
+#include "runtime/methodUnloadBlocker.hpp"
+
+#include "memory/universe.hpp"
+#include "oops/klass.inline.hpp"
+#include "oops/oopHandle.inline.hpp"
+#include "oops/weakHandle.inline.hpp"
+
+inline MethodUnloadBlocker::MethodUnloadBlocker(Method* method) {
+  _method = method;
+  oop obj = get_unload_blocker(method);
+  if (obj != nullptr) {
+    _weak_handle = WeakHandle(Universe::vm_weak(), obj);
+  }
+}
+
+oop MethodUnloadBlocker::get_unload_blocker(Method* method) {
+  assert(method != nullptr, "Should be");
+  InstanceKlass* ik = method->method_holder();
+  oop klass_holder = ik->klass_holder();
+  if (klass_holder != nullptr) {
+    // Normal class, return the holder that would block unloading.
+    // This would be either classloader oop for non-hidden classes,
+    // or Java mirror oop for hidden classes.
+    return klass_holder;
+  }
+
+  // Null holder, the relevant class would not be unloaded.
+  return nullptr;
+}
+
+void MethodUnloadBlocker::release() {
+  _method = nullptr;
+  _weak_handle.release(Universe::vm_weak());
+  _strong_handle.release(Universe::vm_global());
+}
+
+bool MethodUnloadBlocker::is_unloaded() const {
+  // Unloaded if weak handle was set, but now had been cleared by GC.
+  return !_weak_handle.is_empty() && _weak_handle.peek() == nullptr;
+}
+
+inline void MethodUnloadBlocker::block_unloading() {
+  assert(!is_unloaded(), "Pre-condition: should not be unloaded");
+
+  if (!_weak_handle.is_empty()) {
```

Comment:
Does the precondition imply that `!_weak_handle.is_empty()` always hold?

Reply:
Not really. The precondition is:

```
 !is_unloaded() -> 
 !(!_weak_handle.is_empty() && _weak_handle.peek() == nullptr) -> 
 (_weak_handle.is_empty() || _weak_handle.peek() != nullptr)
```

So you see, there is a case when weak handle is empty. It is when `_method` is `nullptr` (default initialized, no method is set), or its unload blocker is `nullptr` (method would be unloaded). Then we bypass the majority of weak->strong dance as unnecessary.